File: backend/app/schemas/transaction.py
from pydantic import BaseModel
from typing import Dict, List, Any, Optional

class InputData(BaseModel):
    ip_address_flag: int = 0
    previous_fraudulent_activity: int = 0
    daily_transaction_count: int = 0
    failed_transaction_count_7d: int = 0
    account_age: int = 0
    transaction_distance: float = 0.0
    is_weekend: int = 0
    is_night: int = 0
    time_since_last_transaction: int = 0
    distance_avg_transaction_7d: float = 0.0
    transaction_to_balance_ratio: float = 0.0
    transaction_amount: float = 0.0
    avg_transaction_amount_7d: float = 0.0
    avg_transaction_distance: float = 0.0
    monthly_transaction_count: int = 0
    transaction_location_flag: int = 0
    suspicious_ip_flag: int = 0
    multiple_account_login: int = 0
    transaction_amount_ratio: float = 0.0
    failed_transaction_rate: float = 0.0
    transaction_distance_ratio: float = 0.0
    transaction_count_ratio: float = 0.0

class RetrainingParams(BaseModel):
    model_id: str
    balanced_ratio: float = 1.5
    version_suffix: Optional[str] = None

File: server/models/transaction.js
import mongoose from 'mongoose';

const transactionSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true }, // who performed the action
  counterparty: { type: mongoose.Schema.Types.ObjectId, ref: 'User' }, // optional
  
  transaction_amount: { type: Number, required: true },
  transaction_type: { type: String, enum: ['DEBIT', 'CREDIT'], required: true },
  description: { type: String },

  status: { type: String, enum: ['SUCCESS', 'PENDING','FAILED'], required: true },

  account_balance: { type: Number }, 
  ip_address: { type: String },
  ip_address_flag: { type: Boolean, default: 0 }, 

  previous_fraudulent_activity: { type: Number, default: 0 }, 
  daily_transaction_count: { type: Number, default: 0 }, 
  avg_transaction_amount_7d: { type: Number, default: 0 }, 
  failed_transaction_count_7d: { type: Number, default: 0 }, 
  account_age: { type: Number }, 

  timestamp: { type: Date, default: Date.now },
  transaction_hour: Number,
  is_weekend: Boolean,
  is_night: { type: Boolean }, 

  time_since_last_transaction: { type: Number, default: 0 }, 
  transaction_to_balance_ratio: { type: Number, default: 0 }, 

  sender_lat: Number,
  sender_long: Number,
  beneficiary_lat: Number,
  beneficiary_long: Number,
  transaction_distance: { type: Number }, // 🆕 Haversine distance between sender & beneficiary
  distance_avg_transaction_7d: { type: Number, default: 0 }, // 🆕 average past distances

  is_fraud: { type: Boolean, default: false }
}, { timestamps: true });


export default mongoose.model('Transaction',transactionSchema );
